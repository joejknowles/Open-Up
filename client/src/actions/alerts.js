export const removeAlert = (id) => ({
  type: 'REMOVE_ALERT',
  id
});
