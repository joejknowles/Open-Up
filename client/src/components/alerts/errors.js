import React from 'react';

export const Errors = ({ errors }) => (
  <div>
    {
      errors.map((error, index) => (
        <p key={ index }>{ error }</p>
      ))
    }
  </div>
);

export default Errors;
