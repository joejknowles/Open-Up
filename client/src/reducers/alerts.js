const isFailureType = ({ type }) => (
  type === 'FETCH_SLOTS_FAILURE' ||
  type === 'BOOK_SLOT_FAILURE'
);

export default (state = [], action) => {
  if (isFailureType(action)) {
    return [
      ...state,
      ...action.response.result
    ];
  }
  switch (action.type) {
    case 'REMOVE_ALERT':
      return [ ...state ].filter((id) => id !== action.id);
    case 'BOOK_SLOT_SUCCESS':
      return [ ...state, action.notificationId ];
    default:
      return state;
  }
};

export const alertsById = (state = {}, action) => {
  if (isFailureType(action)) {
    return {
      ...state,
      ...action.response.errors
    }
  }
  if (action.type === 'BOOK_SLOT_SUCCESS') {
    const newNotification = {
      id: action.notificationId,
      message: 'booked!',
      type: 'SUCCESS'
    };
    return {
      ...state,
      [action.notificationId]: newNotification
    }
  }
  return state;
};
