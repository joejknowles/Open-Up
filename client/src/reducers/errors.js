const isFailureType = ({ type }) => (
  type === 'FETCH_SLOTS_FAILURE' ||
  type === 'BOOK_SLOT_FAILURE'
);

export default (state = [], action) => {
  if (isFailureType(action)) {
    return [
      ...state,
      ...action.response.errors
    ];
  }
  return state;
};
