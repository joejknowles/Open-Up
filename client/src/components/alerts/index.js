import React from 'react';
import { connect } from 'react-redux';
import ReactCSSTransitionsGroup from 'react-addons-css-transition-group';

import '../../styles/Alert.css';

import { alertsSelector, createAlertSelector } from '../../reducers';
import { removeAlert } from '../../actions/alerts'

export const Alert = ({ message, onClick, type }) => (
  <div
    className={ `Alert ${ type === 'ERROR' ? 'Error' : 'Success' }`}
    { ...{ onClick } }>
    { message }
  </div>
);

const mapDispatchToProps = (dispatch, { id }) => ({
  onClick: () => (dispatch(removeAlert(id)))
});
const mapStateToProps = (state, { id }) => createAlertSelector(id);

const ConnectedError = connect(
  mapStateToProps,
  mapDispatchToProps
)(Alert);

export const Alerts = ({ alerts }) => (
  <div className="Alerts">
    <ReactCSSTransitionsGroup
      transitionName="alert"
      transitionEnterTimeout={500}
      transitionLeaveTimeout={400}
      >
      {
        alerts.map((id) => (
          <ConnectedError key={ id } { ...{ id } } />
        ))
      }
    </ReactCSSTransitionsGroup>
  </div>
);

export default connect((state) => (
  { alerts: alertsSelector(state) }
))(Alerts);
