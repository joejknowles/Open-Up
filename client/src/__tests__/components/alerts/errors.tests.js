import { Errors, Error } from '../../../components/alerts/errors';

import React from 'react';
import ReactTestUtils from 'react-addons-test-utils';

const shallowRenderer = ReactTestUtils.createRenderer();

it('errors match snapshot', () => {
  const tree = shallowRenderer.render(
    <Errors alerts={ ['1'] } />
  );
  expect(tree).toMatchSnapshot();
});

it('error matches snapshot', () => {
  const tree = shallowRenderer.render(
    <Error message="error message" />
  );
  expect(tree).toMatchSnapshot();
});
