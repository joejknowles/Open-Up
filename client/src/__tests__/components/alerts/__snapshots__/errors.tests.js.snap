exports[`test error matches snapshot 1`] = `
<div
  className="Alert Error">
  error message
</div>
`;

exports[`test errors match snapshot 1`] = `
<div>
  <Connect(Component)
    id="1" />
</div>
`;
